src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -348,58 +350,52 @@ protected override void Dispose(bool disposing)
                 stateHandle = GCHandle.Alloc(state);
                 requestHandle = CreateRequestHandle(state, stateHandle);
                 state.RequestHandle = requestHandle;
-                needCleanup = true;
-
-                if (state.CancellationToken.IsCancellationRequested)
-                {
-                    state.TrySetCanceled(state.CancellationToken);
-                    return;
-                }
 
-                if (state.RequestMessage.Content != null)
+                if (!state.CancellationToken.IsCancellationRequested && (state.RequestMessage.Content != null))
                 {
                     Stream requestContentStream =
                         await state.RequestMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
-                    if (state.CancellationToken.IsCancellationRequested)
+                    state.RequestContentStream = requestContentStream;
+                    if (!state.CancellationToken.IsCancellationRequested)
```

Comment:
This cancellation check adds value, because it's being done after an await where a cancellation token couldn't be provided to the async operation.  The other two cancellation checks in this method don't seem to add much value, as each of them comes very soon after another check was made.  Doesn't have to be as part of this PR, but I'd suggest removing those extra checks.